### PR TITLE
python3: use _python_sysroot for cross-compilation

### LINKED ISSUE #9275: python3-netifaces: build error caused by linking with -L/usr/lib in configure script
Maintainer: @BKPepe 
Environment: mvebu, arm_cortex-a9+vfpu3, current master

Description:
I couldn't find the bug in the bots, but it fails here consistently.  The failure output is this:
```
if [ "" == "HOST" ]; then export PYTHONPATH="/home/equeiroz/src/openwrt/staging_dir/hostpkg/lib/python3.7:/home/equeiroz/src/openwrt/staging_dir/hostpkg/lib/python3.7/site-packages"; export PYTHONDONTWRITEBYTECODE=0; else export PYTHONPATH="/home/equeiroz/src/openwrt/staging_dir/target-arm_cortex-a9+vfpv3_musl_eabi/usr/lib/python3.7:/home/equeiroz/src/openwrt/staging_dir/target-arm_cortex-a9+vfpv3_musl_eabi//usr/lib/python3.7/site-packages:/home/equeiroz/src/openwrt/build_dir/target-arm_cortex-a9+vfpv3_musl_eabi/netifaces-0.10.9/ipkg-install//usr/lib/python3.7/site-packages"; export PYTHONDONTWRITEBYTECODE=1; export _python_sysroot="/home/equeiroz/src/openwrt/staging_dir/target-arm_cortex-a9+vfpv3_musl_eabi"; export _python_prefix="/usr"; export _python_exec_prefix="/usr"; fi; export PYTHONOPTIMIZE="";   cd /home/equeiroz/src/openwrt/build_dir/target-arm_cortex-a9+vfpv3_musl_eabi/netifaces-0.10.9/; CC="arm-openwrt-linux-muslgnueabi-gcc" CCSHARED="arm-openwrt-linux-muslgnueabi-gcc -fpic" CXX="arm-openwrt-linux-muslgnueabi-g++" LD="arm-openwrt-linux-muslgnueabi-gcc" LDSHARED="arm-openwrt-linux-muslgnueabi-gcc -shared" CFLAGS="-Os -pipe -fno-caller-saves -fno-plt -fhonour-copts -Wno-error=unused-but-set-variable -Wno-error=unused-result -mfloat-abi=hard -iremap/home/equeiroz/src/openwrt/build_dir/target-arm_cortex-a9+vfpv3_musl_eabi/netifaces-0.10.9:netifaces-0.10.9 -Wformat -Werror=format-security -fstack-protector -D_FORTIFY_SOURCE=1 -Wl,-z,now -Wl,-z,relro" CPPFLAGS="-I/home/equeiroz/src/openwrt/staging_dir/target-arm_cortex-a9+vfpv3_musl_eabi/usr/include -I/home/equeiroz/src/openwrt/staging_dir/target-arm_cortex-a9+vfpv3_musl_eabi/include -I/home/equeiroz/src/openwrt/staging_dir/toolchain-arm_cortex-a9+vfpv3_gcc-7.4.0_musl_eabi/usr/include -I/home/equeiroz/src/openwrt/staging_dir/toolchain-arm_cortex-a9+vfpv3_gcc-7.4.0_musl_eabi/include/fortify -I/home/equeiroz/src/openwrt/staging_dir/toolchain-arm_cortex-a9+vfpv3_gcc-7.4.0_musl_eabi/include -I/home/equeiroz/src/openwrt/staging_dir/target-arm_cortex-a9+vfpv3_musl_eabi/usr/include/python3.7" LDFLAGS="-L/home/equeiroz/src/openwrt/staging_dir/target-arm_cortex-a9+vfpv3_musl_eabi/usr/lib -L/home/equeiroz/src/openwrt/staging_dir/target-arm_cortex-a9+vfpv3_musl_eabi/lib -L/home/equeiroz/src/openwrt/staging_dir/toolchain-arm_cortex-a9+vfpv3_gcc-7.4.0_musl_eabi/usr/lib -L/home/equeiroz/src/openwrt/staging_dir/toolchain-arm_cortex-a9+vfpv3_gcc-7.4.0_musl_eabi/lib -znow -zrelro -lpython3.7" _PYTHON_HOST_PLATFORM=linux2 __PYVENV_LAUNCHER__="/usr/bin/python3.7"      /home/equeiroz/src/openwrt/staging_dir/hostpkg/bin/python3.7   ./setup.py   install --prefix="/usr" --root="/home/equeiroz/src/openwrt/build_dir/target-arm_cortex-a9+vfpv3_musl_eabi/netifaces-0.10.9/ipkg-install" --single-version-externally-managed ;
running install
running build
running build_ext
checking for getifaddrs...not found.
checking for getnameinfo...not found.
checking for socket IOCTLs...not found.
checking for optional header files...netpacket/packet.h linux/atm.h linux/llc.h linux/tipc.h linux/dn.h.
checking whether struct sockaddr has a length field...no.
checking which sockaddr_xxx structs are defined...none!
checking for routing socket support...no.
checking for sysctl(CTL_NET...) support...no.
checking for netlink support...no.
building 'netifaces' extension
arm-openwrt-linux-muslgnueabi-gcc -DNDEBUG -Os -pipe -fno-caller-saves -fno-plt -fhonour-copts -Wno-error=unused-but-set-variable -Wno-error=unused-result -mfloat-abi=hard -iremap/home/equeiroz/src/openwrt/build_dir/target-arm_cortex-a9+vfpv3_musl_eabi/Python-3.7.2:Python-3.7.2 -Wformat -Werror=format-security -fstack-protector -D_FORTIFY_SOURCE=1 -Wl,-z,now -Wl,-z,relro -Os -pipe -fno-caller-saves -fno-plt -fhonour-copts -Wno-error=unused-but-set-variable -Wno-error=unused-result -mfloat-abi=hard -iremap/home/equeiroz/src/openwrt/build_dir/target-arm_cortex-a9+vfpv3_musl_eabi/netifaces-0.10.9:netifaces-0.10.9 -Wformat -Werror=format-security -fstack-protector -D_FORTIFY_SOURCE=1 -Wl,-z,now -Wl,-z,relro -I/home/equeiroz/src/openwrt/staging_dir/target-arm_cortex-a9+vfpv3_musl_eabi/usr/include -I/home/equeiroz/src/openwrt/staging_dir/target-arm_cortex-a9+vfpv3_musl_eabi/include -I/home/equeiroz/src/openwrt/staging_dir/toolchain-arm_cortex-a9+vfpv3_gcc-7.4.0_musl_eabi/usr/include -I/home/equeiroz/src/openwrt/staging_dir/toolchain-arm_cortex-a9+vfpv3_gcc-7.4.0_musl_eabi/include/fortify -I/home/equeiroz/src/openwrt/staging_dir/toolchain-arm_cortex-a9+vfpv3_gcc-7.4.0_musl_eabi/include -I/home/equeiroz/src/openwrt/staging_dir/target-arm_cortex-a9+vfpv3_musl_eabi/usr/include/python3.7 -fPIC -DNETIFACES_VERSION=0.10.9 -DHAVE_NETPACKET_PACKET_H=1 -DHAVE_LINUX_ATM_H=1 -DHAVE_LINUX_LLC_H=1 -DHAVE_LINUX_TIPC_H=1 -DHAVE_LINUX_DN_H=1 -I/home/equeiroz/src/openwrt/staging_dir/hostpkg/include/python3.7 -c netifaces.c -o build/temp.linux2-3.7/netifaces.o
cc1: note: someone does not honour COPTS correctly, passed 2 times
netifaces.c:210:6: error: #error You need to add code for your platform.
 #    error You need to add code for your platform.
      ^~~~~
error: command 'arm-openwrt-linux-muslgnueabi-gcc' failed with exit status 1
make[2]: *** [Makefile:47: /home/equeiroz/src/openwrt/build_dir/target-arm_cortex-a9+vfpv3_musl_eabi/netifaces-0.10.9/.built] Error 1
make[2]: Leaving directory '/home/equeiroz/src/openwrt/feeds/packages/lang/python/python3-netifaces'
```
I've checked the build dir, and found out where the problem seems to be:
```
~/src/openwrt/build_dir/target-arm_cortex-a9+vfpv3_musl_eabi/netifaces-0.10.9/build/temp.linux2-3.7 $ cat conftest.out
creating build/temp.linux2-3.7/build
creating build/temp.linux2-3.7/build/temp.linux2-3.7
arm-openwrt-linux-muslgnueabi-gcc -DNDEBUG -Os -pipe -fno-caller-saves -fno-plt -fhonour-copts -Wno-error=unused-but-set-variable -Wno-error=unused-result -mfloat-abi=hard -iremap/home/equeiroz/src/openwrt/build_dir/target-arm_cortex-a9+vfpv3_musl_eabi/Python-3.7.2:Python-3.7.2 -Wformat -Werror=format-security -fstack-protector -D_FORTIFY_SOURCE=1 -Wl,-z,now -Wl,-z,relro -Os -pipe -fno-caller-saves -fno-plt -fhonour-copts -Wno-error=unused-but-set-variable -Wno-error=unused-result -mfloat-abi=hard -iremap/home/equeiroz/src/openwrt/build_dir/target-arm_cortex-a9+vfpv3_musl_eabi/netifaces-0.10.9:netifaces-0.10.9 -Wformat -Werror=format-security -fstack-protector -D_FORTIFY_SOURCE=1 -Wl,-z,now -Wl,-z,relro -I/home/equeiroz/src/openwrt/staging_dir/target-arm_cortex-a9+vfpv3_musl_eabi/usr/include -I/home/equeiroz/src/openwrt/staging_dir/target-arm_cortex-a9+vfpv3_musl_eabi/include -I/home/equeiroz/src/openwrt/staging_dir/toolchain-arm_cortex-a9+vfpv3_gcc-7.4.0_musl_eabi/usr/include -I/home/equeiroz/src/openwrt/staging_dir/toolchain-arm_cortex-a9+vfpv3_gcc-7.4.0_musl_eabi/include/fortify -I/home/equeiroz/src/openwrt/staging_dir/toolchain-arm_cortex-a9+vfpv3_gcc-7.4.0_musl_eabi/include -I/home/equeiroz/src/openwrt/staging_dir/target-arm_cortex-a9+vfpv3_musl_eabi/usr/include/python3.7 -fPIC -I/home/equeiroz/src/openwrt/staging_dir/hostpkg/include/python3.7 -c build/temp.linux2-3.7/conftest-0.c -o build/temp.linux2-3.7/build/temp.linux2-3.7/conftest-0.o
cc1: note: someone does not honour COPTS correctly, passed 2 times
arm-openwrt-linux-muslgnueabi-gcc build/temp.linux2-3.7/build/temp.linux2-3.7/conftest-0.o -L/usr/lib -o build/temp.linux2-3.7/conftest
/home/equeiroz/src/openwrt/staging_dir/toolchain-arm_cortex-a9+vfpv3_gcc-7.4.0_musl_eabi/lib/gcc/arm-openwrt-linux-muslgnueabi/7.4.0/../../../../arm-openwrt-linux-muslgnueabi/bin/ld: skipping incompatible /usr/lib/libc.so when searching for -lc
/home/equeiroz/src/openwrt/staging_dir/toolchain-arm_cortex-a9+vfpv3_gcc-7.4.0_musl_eabi/lib/gcc/arm-openwrt-linux-muslgnueabi/7.4.0/../../../../arm-openwrt-linux-muslgnueabi/bin/ld: /usr/lib/libc.a: error adding symbols: file format not recognized
collect2: error: ld returned 1 exit status
```
If I link the file manually, omitting `-L/usr/lib`, it succeeds; if I _hack_ `TARGET_CC+= $(TARGET_LDFLAGS)` in the Makefile, the package builds, but I don't believe this is a solution, so I'm not opening a PR.  I'm not familiar with python's package build system, so I could not identify where the flag is being added.  @jefferyto, @commodo, do you have any thoughts regarding this?

### PR DESCRIPTION
**Pull request by cotequeiroz**
Maintainer: @commodo @jefferyto 
Compile tested: arm, WRT3200ACM, master - along with python3-netifaces
Run tested: none

Description:
This avoids the use of host header and libraries when building third-party extensions, as done in python2.

Fixes #9275 

Signed-off-by: Eneas U de Queiroz <cotequeiroz@gmail.com>

### CODE REVIEW

**Comment by jefferyto** [comment]
Good catch :joy: can you copy the full patch from Python 2? It would be good to keep the two patches the same (if possible).

File: lang/python/python3/patches/008-distutils-use-python-sysroot.patch [hunk 1]
```diff
@@ -0,0 +1,14 @@
+--- a/Lib/distutils/command/build_ext.py
++++ b/Lib/distutils/command/build_ext.py
+@@ -233,7 +233,10 @@ class build_ext(Command):
+         if (sysconfig.get_config_var('Py_ENABLE_SHARED')):
+             if not sysconfig.python_build:
+                 # building third party extensions
+-                self.library_dirs.append(sysconfig.get_config_var('LIBDIR'))
++                libdir = sysconfig.get_config_var('LIBDIR')
++                if "_python_sysroot" in os.environ:
++                    libdir = os.environ.get("_python_sysroot") + libdir
```
Comment (commodo) [inline lang/python/python3/patches/008-distutils-use-python-sysroot.patch]: maybe use `os.path.join()`
it's a little more robust;
something like:
```
python_sysroot = os.environ.get("_python_sysroot", None)
if (python_sysroot):
   libdir = os.path.join(python_sysroot, libdir)
```

Comment (commodo) [reply lang/python/python3/patches/008-distutils-use-python-sysroot.patch]: I am wondering, if we could adjust`libdir = sysconfig.get_config_var('LIBDIR')` somehow to include `_python_sysroot`;
this looks like a good fix for now, but we may get something similar later

**Comment by commodo** [comment]
rest looks good

File: lang/python/python3/patches/008-distutils-use-python-sysroot.patch (hunk 1 above)
Comment (commodo) [reply lang/python/python3/patches/008-distutils-use-python-sysroot.patch]: ah
nevermind;
saw your comment here https://github.com/openwrt/packages/issues/9275#issuecomment-504563916

this looks fine;
maybe we'll try to upstream this at some point;

**Comment by cotequeiroz** [comment]
> 
> 
> Good catch joy can you copy the full patch from Python 2? It would be good to keep the two patches the same (if possible).

I had just guessed where the problem was--and the first hunk did the job--but you may notice I didn't really ooze much confidence in how these variables are used throughout the system.  Couple that with the addition of the new `BASE_` mirrors not present in python2, used in "virtual" environment, I decided to be minimalist.  A quote from `Doc/library/venv.rst`:
> A virtual environment is a Python environment such that the Python interpreter, libraries and scripts installed into it are isolated from those installed in other virtual environments, and (by default) any libraries installed in a "system" Python, i.e., one which is installed as part of your operating system.
> A virtual environment is a directory tree which contains Python executable files and other files which indicate that it is a virtual environment.
> Common installation tools such as ``Setuptools`` and ``pip`` work as expected with virtual environments. In other words, when a virtual environment is active, they install Python packages into the virtual environment without needing to be told to do so explicitly.

Sounds like this "virtual" environment just could be used to help cross building.

If you would like me to try to apply the rest of the patch, I can easily do it; I guess doing it to `BASE_PREFIX` and `BASE_EXEC_PREFIX` as well.  The latter are used in `get_python_inc` (this seems right), and `get_python_lib`, where I would not be so confident (if it ever gets used to compute a `--rpath` dir, for example).

Another option would be to add a comment to the patch pointing out that the python2 version also patched `Lib/distutils/sysconfig.py` to add `_python_sysroot` to `PREFIX` and `EXEC_PREFIX`, and that the `BASE_` variants did not exist back then.

**Comment by commodo** [comment]
That patch was taken from buildroot.
We may take inspiration from buildroot again if that helps.

**Comment by cotequeiroz** [comment]
They seem to know a lot more than I do.  I'll use  their patch verbatim, just like we did for python2.   I'll push once I'm finished testing it.

**Pull request merged**